File: 147. Insertion Sort List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/*
 * My code is clean and meet the requirement of insertion
 * sort. However, my runtime is 32ms which beats ~42% people
 * on LeetCode. I saw the code whose runtime is 4ms beats 100%.
 * We have the same concept. The only difference is I implement
 * it in a while loop instead of a separate function.
 */
struct ListNode* insertionSortList(struct ListNode* head) {
    struct ListNode dummy, **walk, *rear;
    struct ListNode *cur;

    dummy.next = NULL;
    rear = &dummy;
    cur = head;
    while (cur) {
        head = cur->next;

        /* Insert to dummy list. */
        walk = &dummy.next;
        while (*walk != NULL && (*walk)->val <= cur->val) {
            walk = &(*walk)->next;
        }
        cur->next = *walk;
        *walk = cur;

        cur = head;
    }
    return dummy.next;
}

File: Python3/147. Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        cur = head
        while cur:
            tmp = cur.next
            self.insert(dummy, cur)
            cur = tmp
        return dummy.next

    def insert(self, head: ListNode, node: ListNode) -> None:
        cur = head
        while cur.next and cur.next.val < node.val:
            cur = cur.next
        cur.next, node.next = node, cur.next

    def insertionSortList1(self, head: ListNode) -> ListNode:
        """
        Time Limit Exceeded
        """
        a = []
        cur = head
        while cur:
            a.append(cur.val)
        cur = head
        for item in sorted(a):
            cur.val = item
            cur = cur.next
        return head
